File: sample-objc/ARGEARSample/UI/UIComponent/Permission/PermissionView.m
//
//  AccessAgreeView.m
//  ARGEARSample
//
//  Created by Jihye on 10/09/2019.
//  Copyright © 2019 Seerslab. All rights reserved.
//

#import "PermissionView.h"

@implementation PermissionView

/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect {
    // Drawing code
}
*/

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    if (self) {
    }
    return self;
}

- (void)awakeFromNib {
    [super awakeFromNib];
    
    _permission = [[Permission alloc] init];
    
    [_permissionTitleLabel setText:[@"application_permissions" localized]];
    [_permissionAgreeMessageLabel setText:[@"permission_message" localized]];

    [_permissionDeniedTitleLabel setText:[@"application_denied" localized]];
    [_permissionDeniedMessageLabel setText:[@"denied_message" localized]];

    [_cameraLabel setText:[@"camera" localized]];
    [_storageLabel setText:[@"storage" localized]];
    [_microphoneLabel setText:[@"microphone" localized]];
    
    [_permissionAgreeOKButton setTitle:[@"allow" localized] forState:UIControlStateNormal];
    [_permissionDeniedOKButton setTitle:[@"allow" localized] forState:UIControlStateNormal];
}

- (void)permissionCheck:(PermissionGrantedBlock)granted {
    
    [_permission setGranted:granted];
    
    [self permissionCheck];
}

- (void)permissionCheck {
    [self setStatus:[_permission getPermissionLevel]];
}

- (void)setStatus:(PermissionLevel)status {
    
    _status = status;
    
    switch (status) {
        case PermissionLevelGranted:
            // Hidden
            [_permission granted]();
            [self removeFromSuperview];
            break;
        case PermissionLevelRestricted:
            // show AccessDeniedView
            [_permissionAgreeView setHidden:YES];
            [_permissionDeniedView setHidden:NO];
            break;
        case PermissionLevelNone:
            // show AccessAgreeView
            [_permissionAgreeView setHidden:NO];
            [_permissionDeniedView setHidden:YES];
            break;
        default:
            [self removeFromSuperview];
            break;
    }
}

- (IBAction)permissionAgreeOKButtonAction:(UIButton *)sender {
    
    [_permission permissionAllAllowAction:^{
        [self permissionCheck];
    }];
}

- (IBAction)permissionDeniedOKButtonAction:(UIButton *)sender {

    [_permission openSettings];
}

@end

File: sample-swift/ARGearSample/UI/UIComponent/Permission/PermissionView.swift
//
//  PermissionView.swift
//  ARGearSample
//
//  Created by Jihye on 2020/01/28.
//  Copyright © 2020 Seerslab. All rights reserved.
//

import UIKit

class PermissionView: UIView {
    
    @IBOutlet weak var permissionAgreeView: UIView!
    @IBOutlet weak var permissionTitleLabel: UILabel!
    @IBOutlet weak var permissionAgreeMessageLabel: UILabel!
    
    @IBOutlet weak var permissionDeniedView: UIView!
    @IBOutlet weak var permissionDeniedTitleLabel: UILabel!
    @IBOutlet weak var permissionDeniedMessageLabel: UILabel!

    @IBOutlet weak var permissionAgreeOKButton: UIButton!
    @IBOutlet weak var permissionDeniedOKButton: UIButton!

    @IBOutlet weak var cameraLabel: UILabel!
    @IBOutlet weak var storageLabel: UILabel!
    @IBOutlet weak var microphoneLabel: UILabel!
    
    let permission = Permission()

    override func awakeFromNib() {
        super.awakeFromNib()
        
        self.permissionTitleLabel.text = "application_permissions".localized()
        self.permissionAgreeMessageLabel.text = "permission_message".localized()

        self.permissionDeniedTitleLabel.text = "application_denied".localized()
        self.permissionDeniedMessageLabel.text = "denied_message".localized()

        self.cameraLabel.text = "camera".localized()
        self.storageLabel.text = "storage".localized()
        self.microphoneLabel.text = "microphone".localized()
        
        self.permissionAgreeOKButton.setTitle("allow".localized(), for: .normal)
        self.permissionDeniedOKButton.setTitle("allow".localized(), for: .normal)
    }
    
    @IBAction func permissionAgreeOKButtonAction(_ sender: UIButton) {
        permission.allAllow { permissionLevel in
            self.setPermissionLevel(permissionLevel)
        }
    }
    
    @IBAction func permissionDeniedOKButtonAction(_ sender: UIButton) {
        permission.openSettings()
    }
    
    func setPermissionLevel(_ level: PermissionLevel) {
        
        switch level {
        case .Granted:
            // hide PermissionView
            if let grantedHandler = permission.grantedHandler {
                grantedHandler()
            }
            DispatchQueue.main.async {
                self.removeFromSuperview()
            }
        case .Restricted:
            // show AccessDeniedView
            DispatchQueue.main.async {
                self.permissionAgreeView.isHidden = true
                self.permissionDeniedView.isHidden = false
            }
        case .None:
            // show AccessAgreeView
            DispatchQueue.main.async {
                self.permissionAgreeView.isHidden = false
                self.permissionDeniedView.isHidden = true
            }
        }
    }
}
